go/sync2/consolidator.go

```diff
@@ -25,43 +25,60 @@ import (
 
 // Consolidator consolidates duplicate queries from executing simulaneously
 // and shares results between them.
-type Consolidator struct {
+type Consolidator interface {
+	Create(string) (PendingResult, bool)
+	Items() []ConsolidatorCacheItem
+	Record(query string)
+}
+
+// PendingResult is a wrapper for result of a query.
+type PendingResult interface {
+	Broadcast()
+	Err() error
+	SetErr(error)
+	SetResult(any)
```

Comment:
What's the advantage of using a generic here? Can't this use the `PendingResult` interface? Then you don't have to do those casts and it should be more efficient.

Reply:
Hey @mattlord I think I'm not following the question. My understanding of Golang may be limited here, but I don't think I'm using generics here. I'm just wrapping the current `Result.Result` field in the current codebase with getter/setter fns:

https://github.com/vitessio/vitess/blob/7afc6e1e1816a820baba841fd4dc38901d40e5a3/go/sync2/consolidator.go#L44-L51

That said, I think this _could_ benefit from using generics, so we don't have to do the casting. Happy to introduce that if Vitess team is open to having generic code.

Reply:
`any` is the generic type. That's why you have to cast it to an *sqltypes.Result in places. Why not use the interface type of `PendingResult`?

Reply:
Ah, I glossed over this: 
```
*orig.Result().(*int) 
```

So maybe we do want to support any type as a result? If so, then it's fine.

Reply:
@mattlord is that line you shared from `consolidator_test.go` or from elsewhere? I tried to find where we are using consolidator and it seems like only `query_executor.go` is using it. So if that's right I think we could easily fix the result type to `sqltypes.Result`. 

If we do that, is it OK to import `go/sqltypes` into `go/sync2`? Another thing I can do is make `PendingResult` properly generic, e.g. `PendingResult[T any]`.

Reply:
> @mattlord is that line you shared from `consolidator_test.go` or from elsewhere? I tried to find where we are using consolidator and it seems like only `query_executor.go` is using it. So if that's right I think we could easily fix the result type to `sqltypes.Result`.

Yeah, go/sync2/consolidator_test.go:54

> If we do that, is it OK to import `go/sqltypes` into `go/sync2`? Another thing I can do is make `PendingResult` properly generic, e.g. `PendingResult[T any]`.

If we need to use generics then it's of course fine to do so. But if we don't really need it, we pay a performance penalty and need to worry about casting.

Reply:
From Slack:
 * Going to go ahead an import `sqltypes` into `go/sync2` in order to replace `any` with `sqltypes.Result`.
 * It's possible this would break some undocumented package-dependency rules. Will ask additional person(s) to weigh in on that.